modules/carto/src/api/maps-v3-client.js

```diff
@@ -133,7 +139,6 @@ export async function getData({type, source, connection, credentials, geoColumn,
   log.assert(localCreds.apiVersion === API_VERSIONS.V3, 'Method only available for v3');
   log.assert(localCreds.apiBaseUrl, 'Must define apiBaseUrl');
   log.assert(localCreds.accessToken, 'Must define an accessToken');
-  log.assert(localCreds.mapsUrl, 'mapsUrl cannot be undefined');
 
   if (!localCreds.mapsUrl) {
```

Comment:
Not 100% sure, but I think localCreds.mapsUrl is not overwritten. As we defined mapsUrl at setDefaultCredentials, https://github.com/visgl/deck.gl/blob/master/modules/carto/src/config.js#L52.

Could you try this?

```
setDefaultCredentials({ 
    apiVersion: 'v3',
    apiBaseUrl: 'api1'
})

CartoLayer({
    ...
    credentials: { apiBaseUrl: 'api2'} 
})
```

I think api2 will be ignored

Reply:
I feel that the current behavior of having multiple places where credentials can be set and validates makes the code overcomplicated, especially as `setDefaultCredentials` goes and backfills options which are missing. This makes it hard to see from the outside what is going on and leads to situations which I'm not even sure we can reconcile. For example:

```
setDefaultCredentials({ 
    apiVersion: 'v3',
    apiBaseUrl: 'api1',
    mapsUrl: 'http://custom'
})

CartoLayer({
    ...
    credentials: { apiBaseUrl: 'api2'} 
})
```

It don't think it is clear to the user what is going to take precendence here. Should the explicitly specified `mapsUrl` win? Or should the later `apiBaseUrl: 'api2'` be used to overwrite it? I would say maybe we should take a step back and make `setDefaultCredentials` less clever, and rather validate all the parameters at the point at which we need them. We need to do that anyway as we allow partial overriding of the credentials.

`setDefaultCredentials` then just holds onto whatever is passed to it to help the user simplify this pattern: 

```
// Credentials common to whole app
const defaultCredentials = {...}

// Create layers with optional credential over-rides
CartoLayer({..., credentials: { apiBaseUrl: 'api1', ...defaultCredentials} })
CartoLayer({..., credentials: { apiBaseUrl: 'api2', ...defaultCredentials} })
```

To this:

```
// Credentials common to whole app
const defaultCredentials = {...}
setDefaultCredentials(defaultCredentials);

// Create layers with optional credential over-rides
CartoLayer({..., credentials: { apiBaseUrl: 'api1'} })
CartoLayer({..., credentials: { apiBaseUrl: 'api2} })
```

Reply:
>I would say maybe we should take a step back and make setDefaultCredentials less clever, and rather validate all the parameters at the point at which we need them.
Agree